File: exercises/03 annoying blobs/cs/AnnoyingBlobs/DataBlob.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AnnoyingBlobs
{
    public class DataBlob
    {
        public static DataBlob GetBlob()
        {
            return CreateOrReturnBlob();
        }

        public void AddData(int i)
        {
            data.Add(i);
        }

        public int GetDataAtPos(int p)
        {
            
            return data.ElementAt(p);
        }

        private static DataBlob CreateOrReturnBlob()
        {
            if (theBlob == null)
            {
                Console.WriteLine("Creating Blob");
                theBlob = new DataBlob();
                for (int i = 0; i < 20; i++)
                {
                    theBlob.AddData(i);
                }
            }

            return theBlob;
        }

        private DataBlob()
        {
            data = new List<int>();
        }

        private static DataBlob theBlob = null;
        private IList<int> data;

        public bool IsEmpty()
        {
            return !data.Any();
        }
    }
}

File: solutions/03 annoying blobs/cpp/AnnoyingBlobs/AnnoyingBlobs/DataBlob.cpp
#include "DataBlob.h"
#include <iostream>

using namespace std;

DataBlob* DataBlob::_theBlob = nullptr;

DataBlob& DataBlob::getBlob()
{
	return createOrReturnBlob();
}

void DataBlob::addData(int i)
{
	_data.push_back(i);
}

int DataBlob::getDataAtPos(int p) const
{
	return _data[p];
}

DataBlob::DataBlob()
{
}

DataBlob::~DataBlob()
{
}

DataBlob & DataBlob::createOrReturnBlob()
{
	if (_theBlob == nullptr)
	{
		cout << "Creating Blob" << endl;
		_theBlob = new DataBlob();

		for (int i = 0; i < 20; ++i)
		{
			_theBlob->addData(i);
		}
	}

	return *_theBlob;
}

bool DataBlob::isEmpty() const
{
	return false;
}
